### DISCUSSION TITLE
Error when defining function in GDL on Ubuntu 20

### DISCUSSION THREAD

**Discussion by EdwardEisenhauer** [gnudatalanguage/gdl#1504] [General]
[Stack Overflow question](https://stackoverflow.com/questions/64335330/error-when-defining-function-in-gdl-on-ubuntu-20)

I have a GDL v0.9.9 installed on my machine running Ubuntu 20.04. I am trying to write a function calculating a Julian Date:

```idl
FUNCTION JulianDate, L, M, N
	L1 = L + 4716 - FLOOR((14 - M) / 12)
	M1 = (M + 9) MOD 12
	G = FLOOR(3/4*FLOOR((L1 + 184)/100)) - 38
	RETURN, FLOOR(365.25 * L1) + FLOOR(30.6 * M1 + 0.4) + N - G - 1402
END

JD = JulianDate(1990, 4, 30)
```
Unfortunately, I get the error(s):
```
% Programs can't be compiled from single statement mode.
% Variable is undefined: L
% Execution halted at: $MAIN$          
% Variable is undefined: M
% Execution halted at: $MAIN$          
% Variable is undefined: L1
% Execution halted at: $MAIN$          
% Return statement in procedures cannot have values.
% Parser syntax error: unexpected token: END
% Ambiguous: Variable is undefined: JULIANDATE or: Function not found: JULIANDATE
% Execution halted at: $MAIN$  
```
I compared my code with the [documentation][1] of IDL (as it should be "mutually intelligible") and it looked good for me. Is declaring/defining/calling a function in GDL different than in IDL?

Edit 1: I managed to run my code on the machine with IDL 8.5. The functions must be declared in the beginning of the file before any variables. However it does not change the behaviour of GDL.

  [1]: https://www.harrisgeospatial.com/docs/Defining_a_Function.html

**Comment by alaingdl** [comment]
I am unable to reproduce your problem, with various versions of GDL including git version "around" 0.9.9 

Unclear how do you proceed, the way you show it here is not clear.

A question : did you put the function in a external file ? 

if you run it like : 
```
GDL> @juliandate.pro 
```
it is not OK, and yes you will receive, in IDL or GDL : 
```
% Programs can't be compiled from single statement mode.
```

**Comment by GillesDuvert** [comment]
hello, works for me:
```
IDL> JD = JulianDate(1990, 4, 30)
% Compiled module: JULIANDATE.
IDL> 
$ gdl
GDL> JD = JulianDate(1990, 4, 30)
% Compiled module: JULIANDATE.
```
What you show us is the result of
```
GDL> @ juliandate.pro 
% Programs can't be compiled from single statement mode.
% Variable is undefined: L
% Execution halted at: $MAIN$          
% Variable is undefined: M
% Execution halted at: $MAIN$          
% Variable is undefined: L1
% Execution halted at: $MAIN$          
% Return statement in procedures cannot have values.
```
@ should not be used to call procedures or functions.

note that JULDAY and its reverse CALDAT are already present in GDL:
```
GDL> PRINT, JULDAY(10,4,1582)
     2299160
```

**Comment by EdwardEisenhauer** [comment]
> I am unable to reproduce your problem, with various versions of GDL including git version "around" 0.9.9
> 
> Unclear how do you proceed, the way you show it here is not clear.
> 
> A question : did you put the function in a external file ?
> 
> if you run it like :
> 
> ```
> GDL> @juliandate.pro 
> ```
> 
> it is not OK, and yes you will receive, in IDL or GDL :
> 
> ```
> % Programs can't be compiled from single statement mode.
> ```

All the code is written in a ```list_1``` file and I simply run ```gdl list_1```.

**Comment by GillesDuvert** [comment]
@EdwardEisenhauer IDL behaves the same (below).
in the syntax 'idl list_1' list_1 cannot be a function defintion. It should only contain IDL commands. Perhaps you wanted list_1 to contain only 'JD = JulianDate(1990, 4, 30)' while 'juliandate' should be kept separately in a juliandate.pro file somewhere in the path (itself defined by the !PATH variable). the .pro and lowercaseness of the filename is mandatory.

```
$ idl list_1 
FUNCTION JulianDate, L, M, N
                            ^
% Programs can't be compiled from single statement mode.
  At: /tmp/list_1, Line 1
% Variable is undefined: L.
% Execution halted at: $MAIN$          
% Variable is undefined: M.
% Execution halted at: $MAIN$          
% Variable is undefined: L1.
% Execution halted at: $MAIN$          

        RETURN, FLOOR(365.25 * L1) + FLOOR(30.6 * M1 + 0.4) + N - G - 1402
                                                                          ^
% Return statement in procedures can't have values.
  At: /tmp/list_1, Line 5

END
 ^
% Syntax error.
  At: /tmp/list_1, Line 6
IDL> 
```

**Comment by EdwardEisenhauer** [comment]
Thank you for your clarification. idlde (GUI) runs the same code with no problem (it has to interpret the script differently). Moving the functions into the separate .pro files resolved my issue and code was interpreted correctly by GDL. That you all!

**Comment by GillesDuvert** [comment]
incidentally, there is a [gdlde](https://github.com/gnudatalanguage/gdlde)
